### PR TITLE
Publish collision points for debug purposes

### PR DESCRIPTION
**Pull request by tonynajjar**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | N/A |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Custom Simulation |

---

## Description of contribution in a few bullet points

* Added publish of collision points in form of MarkerArray for debugging purposes

## Description of documentation updates required from your changes

* Docs + Migration guide

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:404]: for some reason it shows up in rviz but not in foxglove, still debugging why

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:81]: Why return? Just store as a class member and grab it here. Lets not complicate the APIs (and would make this not backportable since you change ABI!!)

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:83]:
```suggestion
      "collision_points_marker", 1);
```

Is the same thing

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:403]: I'd remove this so that an empty message is published when empty so that we clear out old stuff

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:406]: ... `now()`?

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:411]: 10cm seems a little big, don't you think?

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:415]: Set to zero so that its only just the most recent message perhaps if we publish the empty message?

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:428]: Remove the second space

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally look good!

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:81]: I would have done it like you proposed as well but that's how it's done for all other parameters of the collision monitor like `cmd_vel_in_topic`, `cmd_vel_out_topic`, `state_topic`. So what do you prefer, breaking ABI or introducing some inconsistency?

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:83]: Not really, `~/collision_points_marker` will result in `/<node_name>/collision_points_marker`. `collision_points_marker` will result in `/collision_points_marker`. Since all other topics are without `~/`, I'll remove it

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:406]: Tried that, for some reason it breaks rviz, which complains about the TF map->baselink requiring extrapolation in the future. My `get_parameter("base_frame_id").as_string()` is `base_link` and my Rviz fixed frame is `map`.

I guess the current implementation is also fine according to [this](https://wiki.ros.org/rviz/DisplayTypes/Marker#:~:text=Note%20that%20the,the%20current%20time.)

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:411]: yep agree, reducing to 0.2

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [reply nav2_collision_monitor/src/collision_monitor_node.cpp]: OK, its just visualization anyway

File: nav2_collision_monitor/src/collision_detector_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_detector_node.cpp:326]: If you check if the publishers have subscriber > 1, then you can remove the parameter and just publish if anyone is listening. We didn't do this historically because that API used to not always work, but it does now. Also for things like the MPPI trajectories or costmaps, we do want to make sure you want them published at all since they're very heavy. This is not

File: nav2_collision_monitor/src/collision_detector_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_detector_node.cpp:328]: unique ptr

File: nav2_collision_monitor/src/collision_detector_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_detector_node.cpp:350]: std::,move(marker_array) once a ptr

This helps with publishing jitter

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Then should be good to go!

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:81]: See new comments below, this may be moot

File: nav2_collision_monitor/test/collision_monitor_node_test.cpp
Comment (tonynajjar) [inline nav2_collision_monitor/test/collision_monitor_node_test.cpp:1234]: fails but not sure why yet

File: nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp
Comment (SteveMacenski) [inline nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:111]: rm

**Review by SteveMacenski — COMMENTED** [review]
Otherwise LGTM once the test works!

File: nav2_collision_monitor/test/collision_monitor_node_test.cpp
Comment (SteveMacenski) [reply nav2_collision_monitor/test/collision_monitor_node_test.cpp:1234]: The points are empty... not sure, you should check!

File: nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp
Comment (SteveMacenski) [reply nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:111]: Ah.... I was never here...

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (tonynajjar) [reply nav2_collision_monitor/src/collision_monitor_node.cpp]: Actually I just ran into a problem with this: I have both a collision detector and monitor running and since they have the same topic name, not having the namespace makes them publish to the same topic. I would add the `~/` if you don't object

**Comment by SteveMacenski** [comment]
@tonynajjar there are multiple collision monitor pkg test failures (ignore python linting… I’ll try to handle that Friday)

**Comment by tonynajjar** [comment]
> @tonynajjar there are multiple collision monitor pkg test failures (ignore python linting… I’ll try to handle that Friday)

Ahh that came for the last topic change, I forgot to do the same in the test

**Comment by tonynajjar** [comment]
Backport to iron please?

**Comment by SteveMacenski** [comment]
beep boop, mergify, do yo thing

**Comment by SteveMacenski** [comment]
the auto backports failed horribly. The two options are for you to wait until I do a manual backport in bulk before a release or you own a PR to do the backport manually yourself. Yours will be done much faster than mine (weeks)

**Pull request merged**
